### ISSUE TITLE
Generated JavaScript Parser Does Not Preserve Left Recursive Expression Subrule Precedences

### ISSUE THREAD

**Issue by derknorton** [antlr/antlr4#2201]
I defined the following simple left recursive expression grammar for antlr4:
```
grammar BugExample;

// Rule Definitions

value: expression EOF ;

real:
    '-'? CONSTANT    #constantReal |
    FLOAT            #variableReal
;

variable: IDENTIFIER ;

expression:  // Precedence (highest to lowest)
    real             #realExpression      |
    variable         #variableExpression  |
//  expression '!'   #factorialExpression |
    '-' expression   #inversionExpression
;


// Token Definitions

FRACTION: '.' ('0'..'9')* '1'..'9' ;

CONSTANT: 'e' | 'pi' ;

FLOAT: INTEGER FRACTION? ('e' INTEGER)? ;

IDENTIFIER: ('a'..'z'|'A'..'Z') ('a'..'z'|'A'..'Z'|'0'..'9')* ;

SPACE: (' '|'\t'|'\r'|'\n')+ -> channel(HIDDEN) ;

fragment
NATURAL: '1'..'9' ('0'..'9')* ;

fragment
INTEGER: '0' | '-'? NATURAL ;
```
Notice the commented out factorial expression in the expression types. Also notice that the definition of a FLOAT token allows for negative values so a negative real expression should take precedence over an inversion expression. With the factorial expression commented out, the generated JS parser does indeed parse a negative constant '-e' correctly as a real expression. However, if we uncomment the factorial expression (which should not have anything to do with the test example) and regenerate the parser, the '-e' is suddenly parsed as an inversion expression. Here is the test code showing it:
```
'use strict';

var language = require('../BugExample');
var testCase = require('nodeunit').testCase;

module.exports = testCase({
    'Test Parser': function(test) {
        var testValues = ['5.27e-15', '-5.3e22','e', '-e', 'expo', '-expo'];
        var expectedResults = [
            'RealExpressionContext',      // positive real number
            'RealExpressionContext',      // negative real number
            'RealExpressionContext',      // positive real constant
            'RealExpressionContext',      // positive real constant
            'VariableExpressionContext',  // variable value
            'InversionExpressionContext'  // negative variable value
        ];
        test.expect(testValues.length);
        for (var i = 0; i < testValues.length; i++) {
            var value = testValues[i];
            console.log('\nTesting: ' + value);
            var expression = language.parseValue(value).getChild(0);
            test.strictEqual(expression.constructor.name, expectedResults[i]);
        }
        test.done();
    }
});
```
It turns out that adding any left recursive sub-rule type listed in the "Definitive Antlr 4 Reference" that begins with the expression e.g. "binary", "ternary", and "unary suffix" expressions, will cause this problem.  However, the problem does **not** show up in the generated Java parser.

I created a GitHub project showing the problem with both JS and Java unit tests to make it easier to experiment and debug: https://github.com/derknorton/antlr4-bug-example

When I compare the generated parsers for the JS and Java bindings I found that the Java parser maintains the precedence ordering of the expression sub-rule case blocks in the ExpressionContext:expression() method, but the JS parser randomizes the order of the case blocks once any left recursive sub-rule starting with the expression is added to the expression rule.

I suspect this is the problem but the code is too complex for my poor brain ;-)  I need an antlr guru to determine this for sure.  I did not check any other language bindings so I can't offer any information about them.

This issue was first raised here: https://stackoverflow.com/questions/48195702/my-antlr-4-left-recursive-expression-not-preserving-the-expression-precedence-or

This is a fairly serious issue for me since I am tied to JS and now will need to restructure my grammar to use the old (antlr 3) style expression format which seems a shame.

**Comment by ericvergnaud** [comment]
Hi,
I’ll try and look into this over the week end.
Eric 

Envoyé de mon iPhone

**Comment by derknorton** [comment]
Thanks Eric, I really appreciate it!

Derk

**Comment by ericvergnaud** [comment]
Derk,

not sure what I'm doing wrong but I fail to reproduce the issue.
In my branch: https://github.com/ericvergnaud/antlr4/tree/explore-javascript-issue
I have added a test in LeftRecursionDescriptors with the above grammar and uncommented factorial expression.
Both java and javascript generate a similar 'expression' parser, and when executed using 'value' as start rule successfully treat input "-e" as a 'realExpression'.
I'm attaching the generated parsers for reference, so you can compare with yours.

[BaseNodeTest-main-1515811128800.zip](https://github.com/antlr/antlr4/files/1628004/BaseNodeTest-main-1515811128800.zip)
[BaseJavaTest-main-562293174117958.zip](https://github.com/antlr/antlr4/files/1628005/BaseJavaTest-main-562293174117958.zip)

**Comment by ericvergnaud** [comment]
I would add that you can certainly work around the issue if any by:
 - putting lexer rules in a dedicated lexer grammar so you can control precedence accurately
 - handling negative numbers and constants in the lexer rather than the parser

**Comment by ericvergnaud** [comment]
I also notice that your JS parser was generated using 4.5.1:
// Generated from grammar/BugExample.g4 by ANTLR 4.5.1

Current version is 4.7.1 and possibly fixes an unknown bug

**Comment by derknorton** [comment]
Wow, you are right.  I was assuming that since my grunt package.json file references version 4.7.1 of antlr4 that the generation would use that version but it appears that the latest version of the grunt-antlr4 plugin (0.1.6) is only on version 4.5.1.  I will take a look at your fork and see how you got it to run.  Thanks so much for your quick response Eric!

**Comment by derknorton** [comment]
Eric is exactly right.  This was an unknown bug in version 4.5.1 of antlr4 and has been fixed by version 4.7.1.  I verified that the case blocks in the expression() method preserve the order of precedence of the sub-rules.  Thanks again Eric for your time!
